README.md
# Github-CCTray

A lightweight App that provides [all Github Workflows of a Repository](https://docs.github.com/en/rest/actions/workflow-runs?apiVersion=2022-11-28#list-workflow-runs-for-a-repository) in [CCTray Specification](https://cctray.org/v1/).

![STATUS](https://github.com/mansab/github-cctray/actions/workflows/.github/workflows/pylint.yml/badge.svg)
![STATUS](https://github.com/mansab/github-cctray/actions/workflows/.github/workflows/unittests.yml/badge.svg)
![STATUS](https://github.com/mansab/github-cctray/actions/workflows/.github/workflows/codeql.yml/badge.svg)

# Usecase
Visibility for your Github Actions & Workflows in one place.

You can use the App to configure [CCTray Clients](https://cctray.org/clients/):
* [CCMenu](https://ccmenu.org/) `[tested]`
* [Nevergreen Dashboard](https://github.com/build-canaries/nevergreen) `[tested]`

<img src="https://github.com/mansab/github-cctray/raw/main/images/ccmenu_desktop_menu_bar.png?raw=true" width=40% height=50%> <img src="https://github.com/mansab/github-cctray/blob/main/images/ccmenu_projects_configuration.png?raw=true" width=40% height=50%>
<img src="https://github.com/mansab/github-cctray/blob/main/images/nevergreen_dashboard.png?raw=true" width=40% height=50%> <img src="https://github.com/mansab/github-cctray/blob/main/images/nevergreen_dashboard_configuration.png?raw=true" width=40% height=50%>

# Running the App

## Prerequisites

### Authentication Token

To authenticate with Github API, the app needs a `token`, 
it can be provided by **either** of the following methods:

#### Github Personal Access Token
* [FGPAT (recommended) or PAT](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token)
* Read-Only access to Actions (Workflows, workflow runs and artifacts) required for Private repos.
* You will need to set: `GITHUB_TOKEN="<your_token>"` as the environment variable.

#### Github APP (Recommened)
* [Create a Github APP](https://docs.github.com/en/apps/creating-github-apps/registering-a-github-app/registering-a-github-app) and grant Read-Only access to Github Actions
* The Github APP should be installed on the Github Organization or Account with access to the required repositories.
* You will need to set the following environment variables:
```
APP_AUTH_ID=<id_of_your_github_app>
APP_AUTH_PRIVATE_KEY_B64=<private_key_of_your_github_app_bas64_encoded>
APP_AUTH_INSTALLATION_ID=<installtion_id_once_installed>
```
* Please refer to Github's offical documentation to know what these values are and where can you find them 

**Please take into account the [Github API rate limit](https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#rate-limiting) for authentication tokens.**

## With Docker

### Build the Docker image

```bash
docker build -t github-cctray:latest . 
```

### Launch the Docker container

You can do this in two ways:

* Personal Access Token method

```bash
 docker run -p 8000:8000 \
            -e GITHUB_TOKEN="<your_token>" \
            -e BASIC_AUTH_USERNAME="<your_username>" \
            -e BASIC_AUTH_PASSWORD="<your_password>" \
            github-cctray:latest
```

* Github App method

```bash
 docker run -p 8000:8000 \
            -e APP_AUTH_ID="<id_of_your_github_app>" \
            -e APP_AUTH_PRIVATE_KEY_B64="<private_key_of_your_github_app_base64_encoded>" \
            -e APP_AUTH_INSTALLATION_ID="<installtion_id_once_installed>" \
            -e BASIC_AUTH_USERNAME="<your_username>" \
            -e BASIC_AUTH_PASSWORD="<your_password>" \
            github-cctray:latest --mode app-auth
```

# Usage

Once up, the App binds to port `8000` by default and should be available at: http://localhost:8000

## Making an HTTP request

The App accepts GET requests with following parameters:

**manadatory parameters**

* `owner` - Organisation or User who owns the repository
* `repo` - Name of the Repository

**optional parameter**

* `token` - If you want to use FGPAT per user to access the API, to overcome Github API rate limiting (this takes precedence over the token/Github App auth set in the env var).

For Example:

* Mandatory Parameters
```bash
curl -X GET http://localhost:8000?owner=<repo_owner>&repo=<repository_name>
```

* Optional Parameter
```bash
curl -X GET http://localhost:8000?owner=<repo_owner>&repo=<repository_name&token=<your_token>
```

## Response

The above request would return an XML response (CCTray Specification) with all the workflows of a repository, while filtering them to return only **Unique & Latest** runs:

```bash
<Projects>
<Project name="github-cctray/CodeQL" activity="Sleeping" lastBuildStatus="Success" lastBuildTime="2023-05-07T23:22:02Z" webUrl="https://github.com/mansab/github-cctray/actions/runs/4909813101"/>
<Project name="github-cctray/Pylint" activity="Sleeping" lastBuildStatus="Success" lastBuildTime="2023-05-07T23:19:13Z" webUrl="https://github.com/mansab/github-cctray/actions/runs/4909813107"/>
<Project name="github-cctray/TestUnit" activity="Sleeping" lastBuildStatus="Success" lastBuildTime="2023-05-07T23:18:59Z" webUrl="https://github.com/mansab/github-cctray/actions/runs/4909813102"/>
</Projects>
```

Attributes are returned as follows:

<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Description</th>
      <th>Type</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>name</td>
      <td>Name of the repository and the workflow</td>
      <td>string : repo_name/workflow</td>
    </tr>
    <tr>
      <td>activity</td>
      <td>the current state of the project</td>
      <td>string enum : Sleeping, Building</td>
    </tr>
    <tr>
      <td>lastBuildStatus</td>
      <td>a brief description of the last build</td>
      <td>string enum : Success, Failure, Unknown</td>
    </tr>
    <tr>
      <td>lastBuildTime</td>
      <td>when the last build occurred</td>
      <td>DateTime</td>
    </tr>
    <tr>
      <td>webUrl</td>
      <td>Exact URL of the Github Action run for a workflow</td>
      <td>string (URL)</td>
    </tr>
    <tr>
     <td>lastBuildLabel</td>
     <td>Short commit ID associated with the last workflow run</td>
     <td>string : Git SHA (8 char)</td>
    </tr>
  </tbody>
</table>

## Health check

The App has an health endpoint which returns the status and version of the app

```bash
curl -X GET http://localhost:8000/health
```


### Response

```
{"status":"ok","version":"2.2.0"}
```

## Rate Limiting

Github has Rate Limiting for their APIs, to check if you are rate limited, use this endpoint

* With token in the environment variable
```bash
curl -X GET http://localhost:8000/limit
```

* With token in the query parameter
```bash
curl -X GET http://localhost:8000/limit?token=<your_token>
```

### Response

```
{"rate_limit":{"limit":5000,"remaining":1724,"reset":1686920826,"reset_cest":"2023-06-16 15:07:06 UTC+02:00+0200","used":3276},"status":"ok"}
```

# Development Setup

* Python 3.9
* pip 
* Activate [Python virtualenv](https://python.land/virtual-environments/virtualenv)

```bash
python -m venv venv
source venv/bin/activate
```

* Install requirements

```bash
pip install -r requirements.txt
```

* Execute

```bash
* set necessary env variable to authenticate with Github (see Prerequisites)
* export BASIC_AUTH_USERNAME=<user>
* export BASIC_AUTH_PASSWORD=<pass>
* python app.py --mode [pat-auth|app-auth] # pat-auth is the default mode if no mode is set
```



CHANGELOG.md
## 3.2.0

* Expect a Base64 encoded private key of Github App for consistency
* Improve logs

## 3.1.1

* Bug fix for the /health route

## 3.1.0

* Modularize the source code
* Create separate modules for app, helpers, routes & config
* Update unit tests

## 3.0.1

* Fix for v3.0.0
* Update the Dockerfile
* Allow overriding the python command attribute to switch b/w pat-auth and app-auth
* Update readme

## 3.0.0

* Add support for Github App based authentication
* Return git commit ID as buildlabel in the XML

## 2.2.0

Added /health endpoint

## 2.1.0

Added Authentication functionality

## 2.0.0

Multi threading and improved performance

## 1.0.0

Initial Release


tests/test_app.py
import unittest
from unittest.mock import patch
from src.routes import app

class RoutesTestCase(unittest.TestCase):

    def setUp(self):
        """Set up the test environment."""
        app.testing = True
        app.config['BASIC_AUTH_USERNAME'] = 'user'
        app.config['BASIC_AUTH_PASSWORD'] = 'pass'
        self.client = app.test_client()
        self.headers = {
            'Authorization': 'Basic dXNlcjpwYXNz'
        }

    @patch("src.routes.get_token")
    @patch("src.routes.get_all_workflow_runs")
    def test_index(self, mock_get_all_workflow_runs, mock_get_token):
        """Test case for the index route."""
        mock_get_token.return_value = "mocked_token"

        # Mock the response from get_all_workflow_runs
        mock_get_all_workflow_runs.return_value = [
            {
                "name": "workflow1",
                "status": "completed",
                "conclusion": "success",
                "updated_at": "2023-07-05T12:00:00Z",
                "html_url": "https://github.com/test_owner/test_repo/actions/runs/1",
                "head_commit": {"id": "1234567890abcdef"}
            },
            {
                "name": "workflow2",
                "status": "completed",
                "conclusion": "failure",
                "updated_at": "2023-07-04T12:00:00Z",
                "html_url": "https://github.com/test_owner/test_repo/actions/runs/2",
                "head_commit": {"id": "abcdef1234567890"}
            }
        ]

        response = self.client.get('/?owner=test_owner&repo=test_repo', headers=self.headers)
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.headers['Content-Type'], 'application/xml')

        expected_xml = """<Projects>
        <Project name="test_repo/workflow1" activity="Sleeping" lastBuildStatus="Success" lastBuildTime="2023-07-05T12:00:00Z" webUrl="https://github.com/test_owner/test_repo/actions/runs/1" lastBuildLabel="12345678" />
        <Project name="test_repo/workflow2" activity="Sleeping" lastBuildStatus="Failure" lastBuildTime="2023-07-04T12:00:00Z" webUrl="https://github.com/test_owner/test_repo/actions/runs/2" lastBuildLabel="abcdef12" />
        </Projects>""".replace('\n', '').replace(' '*8, '')

        self.maxDiff = None  # Set self.maxDiff to None
        self.assertEqual(response.data.decode(), expected_xml)

        mock_get_token.assert_called_once()
        mock_get_all_workflow_runs.assert_called_once_with("test_owner", "test_repo", "mocked_token")


if __name__ == '__main__':
    unittest.main()


cctray.py
import os
import requests as req
import xml.etree.ElementTree as ET
from flask import Flask, request, make_response


app = Flask('github-cctray')

def get_workflow_runs(owner, repo, token):
    endpoint = f"https://api.github.com/repos/{owner}/{repo}/actions/runs"
    headers = {
        "Authorization": f"Bearer {token}",
        "Accept": "application/vnd.github.v3+json"
    }
    response = req.get(endpoint, headers=headers)
    return response.json()

@app.route('/')
def index():
    owner = request.args.get("owner") or request.form.get('owner')
    repo = request.args.get("repo") or request.form.get('repo')
    token = os.environ.get("GITHUB_TOKEN")

    if not owner or not repo or not token:
        return make_response("Missing parameters", 400)

    data = get_workflow_runs(owner, repo, token)

    root = ET.Element("Projects")

    for run in data["workflow_runs"]:
        project = ET.SubElement(root, "Project")

        project.set("name", repo)

        if run["conclusion"] == "success":
            project.set("lastBuildStatus", "Success")
        elif run["conclusion"] == "failure":
            project.set("lastBuildStatus", "Failure")
        else:
            project.set("lastBuildStatus", "Unknown")

        project.set("lastBuildTime", run["updated_at"])

        project.set("webUrl", run["html_url"])

    response = make_response(ET.tostring(root).decode())
    response.headers['Content-Type'] = 'application/xml'
    return response

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=8000)


cctray-2.py
import os
import requests
import xml.etree.ElementTree as ET
from datetime import datetime

owner = os.environ.get("GITHUB_USERNAME")
repo = os.environ.get("GITHUB_REPOSITORY")
token = os.environ.get("GITHUB_TOKEN")

# GitHub API endpoint to retrieve workflow runs for a repository
endpoint = "https://api.github.com/repos/{owner}/{repo}/actions/runs"

# Build the API URL
url = endpoint.format(owner=owner, repo=repo)

# Add authentication to the request headers
headers = {
    "Authorization": f"Bearer {token}",
    "Accept": "application/vnd.github.v3+json"
}

# Send the API request
response = requests.get(url, headers=headers)

# Parse the response JSON
runs = response.json()["workflow_runs"]

# Build the CCTray-compatible XML response
root = ET.Element("Projects")

for run in runs:
    project = ET.SubElement(root, "Project")
    project.set("name", run["name"])
    project.set("lastBuildStatus", "Unknown")
    project.set("lastBuildLabel", run["head_branch"])
    project.set("activity", "Building" if run["status"] == "in_progress" else "Sleeping")
    project.set("webUrl", run["html_url"])
    project.set("lastBuildTime", datetime.strptime(run["created_at"], "%Y-%m-%dT%H:%M:%SZ").strftime("%Y-%m-%dT%H:%M:%S.%fZ"))

# Output the XML response
xmlstr = ET.tostring(root, encoding="unicode")
print(xmlstr)



src/helpers.py
"""Helpers Module"""
import datetime
import re
import base64
import argparse
from concurrent.futures import ThreadPoolExecutor
import jwt
import requests
from flask import request
from config import (
    API_BASE_URL,
    MAX_WORKERS,
    TIMEOUT,
    GITHUB_TOKEN,
    APP_AUTH_ID,
    APP_AUTH_BASE_URL,
    APP_AUTH_INSTALLATION_ID,
    APP_AUTH_PRIVATE_KEY_B64,
)


def decode_base64(value):
    """Decodes a Base64 string

    Returns:
        value: Base64 Decoded value, utf-8 format.
    """
    return base64.b64decode(value).decode("utf-8")


def redact_token(uri):
    """Redacts the 'token' value from the URI

    Returns:
        uri: token=<REDACTED> 
    """
    return re.sub(r'(\?|&)token=.*?(&|$)', r'\1token=<REDACTED>\2', uri)


def get_token():
    """Sets the GitHub API token based on the selected mode

    Returns:
        token: Either the personal access token or the GitHub App access token
    """
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--mode",
        choices=["pat-auth", "app-auth"],
        default="pat-auth",
        help="Authentication mode",
    )
    args = parser.parse_args()

    token = request.args.get("token")

    if token:
        return token
    elif args.mode == "pat-auth":
        token = GITHUB_TOKEN
    elif args.mode == "app-auth":
        app_auth_id = APP_AUTH_ID
        app_auth_private_key_b64 = decode_base64(APP_AUTH_PRIVATE_KEY_B64)
        app_auth_installation_id = APP_AUTH_INSTALLATION_ID
        app_auth_base_url = APP_AUTH_BASE_URL

        now = datetime.datetime.utcnow()
        iat = int((now - datetime.datetime(1970, 1, 1)).total_seconds())
        exp = iat + 600
        payload = {"iat": iat, "exp": exp, "iss": app_auth_id}
        encoded_jwt = jwt.encode(payload, app_auth_private_key_b64, algorithm="RS256")
        headers = {
            "Authorization": f"Bearer {encoded_jwt}",
            "Accept": "application/vnd.github.v3+json",
        }
        response = requests.post(
            f"{app_auth_base_url}/app/installations/{app_auth_installation_id}/access_tokens",
            headers=headers,
            timeout=TIMEOUT,
        )

        if response.status_code == 201:
            token = response.json()["token"]
        else:
            raise Exception(
                f"Failed to obtain access token: {response.status_code} {response.text}"
            )

    return token


def get_workflows(owner, repo, headers):
    """Get the workflows for a given owner and repo from the GitHub API.

    Args:
        owner (str): The owner of the repository.
        repo (str): The repository name.
        headers (dict): HTTP headers to be sent with the request.

    Returns:
        list: A list of workflows for the given repository.

    Raises:
        requests.HTTPError: If the request to the GitHub API fails.
    """
    endpoint = f"{API_BASE_URL}/repos/{owner}/{repo}/actions/workflows"
    response = requests.get(endpoint, headers=headers, timeout=TIMEOUT)
    response.raise_for_status()
    return response.json()["workflows"]


def get_workflow_runs(workflow, headers):
    """Get the workflow runs for a specific workflow from the GitHub API.

    Args:
        workflow (dict): The workflow information.
        headers (dict): HTTP headers to be sent with the request.

    Returns:
        list: A list of workflow runs for the given workflow.

    Raises:
        requests.HTTPError: If the request to the GitHub API fails.
    """
    url = f"{workflow['url']}/runs"
    response = requests.get(url, headers=headers, timeout=TIMEOUT)
    response.raise_for_status()
    return response.json()["workflow_runs"]


def get_all_workflow_runs(owner, repo, token):
    """Get all workflow runs for a given owner, repo, and token.

    Args:
        owner (str): The owner of the repository.
        repo (str): The repository name.
        token (str): The GitHub token for authentication.

    Returns:
        list: A list of all workflow runs for the given repository.

    Raises:
        requests.HTTPError: If the request to the GitHub API fails.
    """
    headers = {
        "Authorization": f"Bearer {token}",
        "Accept": "application/vnd.github.v3+json",
    }

    workflows = get_workflows(owner, repo, headers)
    with ThreadPoolExecutor(max_workers=MAX_WORKERS) as executor:
        futures = [
            executor.submit(get_workflow_runs, workflow, headers)
            for workflow in workflows
        ]

    results = []
    for future in futures:
        data = future.result()
        results += data

    return results


src/routes.py
"""Routes Module"""
import re
import logging
import xml.etree.ElementTree as ET
import datetime
import requests
from flask import Flask, request, make_response, jsonify
from flask_basicauth import BasicAuth
from helpers import get_token, get_all_workflow_runs, redact_token
from config import BASIC_AUTH_USERNAME, BASIC_AUTH_PASSWORD, TIMEOUT

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

app = Flask('github-cctray')
app.config['BASIC_AUTH_USERNAME'] = BASIC_AUTH_USERNAME
app.config['BASIC_AUTH_PASSWORD'] = BASIC_AUTH_PASSWORD

basic_auth = BasicAuth(app)

@app.route('/')
@basic_auth.required
def index():
    """Endpoint for generating the CCTray XML.

    Returns:
        flask.Response: The XML response containing the project information.
    """
    owner = request.args.get("owner") or request.form.get('owner')
    repo = request.args.get("repo") or request.form.get('repo')
    token = get_token()

    if not owner or not repo or not token:
        logger.warning("Missing parameter(s) or Environment Variable")
        return make_response("Missing parameter(s)", 400)

    data = get_all_workflow_runs(owner, repo, token)

    workflow_runs = sorted(
        data,
        key=lambda run: run["updated_at"],
        reverse=True)

    root = ET.Element("Projects")
    project_names = set()  # Set to store unique project names

    for run in workflow_runs:
        project_name = repo + "/" + run["name"]
        if project_name not in project_names:  # Check if project name is already in the set
            project_names.add(project_name)  # Add project name to the set
            project = ET.SubElement(root, "Project")
            project.set("name", project_name)

            # Map 'Github' attributes to 'CCTray'
            if run["status"] == "completed":
                if run["conclusion"] == "success":
                    project.set("activity", "Sleeping")
                elif run["conclusion"] == "failure":
                    project.set("activity", "Sleeping")
            else:
                project.set("activity", "Building")

            project.set("lastBuildStatus", "Success"
                        if run["conclusion"] == "success"
                        else "Failure"
                        if run["conclusion"] == "failure"
                        else "Unknown")
            project.set("lastBuildTime", run["updated_at"])
            project.set("webUrl", run["html_url"])
            short_commit_id = run["head_commit"]["id"][:8]
            project.set("lastBuildLabel", short_commit_id)

    response = make_response(ET.tostring(root).decode())
    response.headers['Content-Type'] = 'application/xml'

    logger.info("Request URI: %s Response Code: %d",
                redact_token(request.full_path), response.status_code)

    return response


@app.route('/health')
def health():
    """Endpoint for checking the health status.

    Returns:
        flask.Response: JSON response containing the status and version.
    """
    with open('../CHANGELOG.md', 'r', encoding='utf-8') as changelog_file:
        changelog_content = changelog_file.read()

    latest_version_match = re.search(
        r'##\s*(\d+\.\d+\.\d+)', changelog_content)
    latest_version = latest_version_match.group(
        1) if latest_version_match else 'Unknown'

    response = {
        'status': 'ok',
        'version': f'{latest_version}'
    }

    return jsonify(response)


@app.route('/limit')
@basic_auth.required
def limit():
    """Endpoint for checking the rate limit status.

    Returns:
        flask.Response: JSON response containing rate limiting information.
    """
    token = get_token()
    headers = {
        'Accept': 'application/vnd.github+json',
        "Authorization": f"Bearer {token}",
        'X-GitHub-Api-Version': '2022-11-28'
    }
    url = 'https://api.github.com/rate_limit'
    response = requests.get(url, headers=headers, timeout=TIMEOUT)

    if response.status_code == 200:
        rate = response.json().get('rate', {})
        reset_unix_time = rate.get('reset', 0)
        reset_datetime = datetime.datetime.fromtimestamp(reset_unix_time)
        reset_cest = reset_datetime.astimezone(
            datetime.timezone(datetime.timedelta(hours=2)))
        rate['reset_cest'] = reset_cest.strftime('%Y-%m-%d %H:%M:%S %Z%z')

        if rate.get('remaining', 0) == 0:
            response = {
                'status': 'rate_limited',
                'rate_limit': rate
            }
        else:
            response = {
                'status': 'ok',
                'rate_limit': rate
            }
    else:
        response = {'status': 'ok', 'rate_limit': {
            'error': 'Failed to retrieve rate limit information'}}

    return jsonify(response)


@app.errorhandler(Exception)
def handle_error(exception):
    """Error handler for handling exceptions raised in the application.

    Args:
        exception (Exception): The exception object.

    Returns:
        str: The error message response.
    """
    logger.error("An error occurred: %s Request URI: %s", str(exception),
                 redact_token(request.full_path))
    error_message = f"An error occurred: {str(exception)}"
    return error_message, 500
